crates/bin/pd/src/compat.rs

```diff
@@ -0,0 +1,122 @@
+//! Logic for determining protocol compatibility, as defined by the
+//! workspace-wide APP_VERSION value. On startup, `pd` will inspect
+//! the APP_VERSION set in the local non-verifiable storage, and warn
+//! or fail about mismatches.
+//!
+//!
+//! TODO: move this code into the `app` crate.
+use anyhow;
+use cnidarium::{StateDelta, StateRead, StateWrite, Storage};
+// use penumbra_app::app::StateReadExt as _;
+use penumbra_app::APP_VERSION;
+
+const APP_VERSION_KEY: &str = "app_version";
+
+/// Retrieve the APP_VERSION last written to local non-verifiable storage.
+/// Returns an Option, because as late as v0.80.8, APP_VERSION was not
+/// written to local state.
+pub async fn get_state_version(storage: Storage) -> anyhow::Result<Option<u64>> {
+    let snapshot = storage.latest_snapshot();
+    let result = snapshot
+        .nonverifiable_get_raw(APP_VERSION_KEY.as_bytes())
+        .await?;
+
+    let local_app_version = match result {
+        Some(v) => {
+            let app_version: AppVersion = v.try_into()?;
+            Some(app_version.try_into()?)
+        }
+        None => None,
+    };
+    Ok(local_app_version)
+}
+
+/// Check whether the local state matches the current APP_VERSION.
+/// If it's not set yet, set it to the current value.
+/// If it's a surprising value, error.
+pub async fn check_state_version(storage: Storage) -> anyhow::Result<u64> {
+    let local_version = get_state_version(storage.clone()).await?;
+    match local_version {
+        Some(v) => {
+            if v > APP_VERSION {
+                anyhow::bail!(
+                    "state version {v} is newer than current app version {APP_VERSION}; you should upgrade pd",
+                )
+            } else if v < APP_VERSION {
+                anyhow::bail!(
+                    "state version {v} is older than current app version {APP_VERSION}; you should run 'pd migrate'",
+                )
+            } else {
+                Ok(v)
+            }
+        }
+        // If not set, set it.
+        None => {
+            set_state_version(storage, APP_VERSION).await?;
+            Ok(APP_VERSION)
+        }
+    }
+}
+
+/// Write the given version to the local non-verifiable storage.
+pub async fn set_state_version(storage: Storage, version: u64) -> anyhow::Result<()> {
+    let v = AppVersion(version);
+    let snapshot = storage.latest_snapshot();
+    let mut delta = StateDelta::new(snapshot);
+    delta.nonverifiable_put_raw(APP_VERSION_KEY.to_string().into(), v.try_into()?);
+    Ok(())
+}
+
+/// Wrapper struct representing the `APP_VERSION`, intended
+/// for custom TryInto/TryFrom implementations.
+#[derive(Debug)]
+struct AppVersion(u64);
```

Comment:
If we write a proto to storage we can dispense with the wrapper type and custom serialization. In general, we should try to avoid doing custom serialization and using the `_raw` methods, preferring extension traits like `StateWriteProto`.

Here's an example of writing a primitive type to NV storage: https://github.com/penumbra-zone/penumbra-security-e/blob/12a2e7fff02414a0292ad9da50f90fd2f2724467/crates/core/component/governance/src/component/view.rs#L978